### PR TITLE
TC: by_name functions for projects and datasets

### PR DESCRIPTION
**Pull request by skalish**
# ↪️ Pull Request

Adds functions for getting Tamr projects and datasets by name

## 💻 Examples

```python
import tamr_client as tc

auth = tc.UsernamePasswordAuth(...)
s = tc.session.from_auth(auth)
instance = tc.Instance("http", "localhost", 9100)

dataset = tc.dataset.by_name(s, instance, "My Dataset")
project = tc.project.by_name(s, instance, "My Project")
```

## ✔️ PR Todo

- [x] Added/updated testing for this change
- [ ] Included links to related issues/PRs
- [x] Update relevant [docs](https://github.com/Datatamer/tamr-client/tree/master/docs) + docstrings
- [x] Update the [CHANGELOG](https://github.com/Datatamer/tamr-client/blob/master/CHANGELOG.md) under the current `-dev` version:
  - Add changelog entries under any that apply: **BREAKING CHANGES**, **NEW FEATURES**, **BUG FIXES**.
  - Changelog entry format: `[#<issue number>](<link to issue>) <change description>`

### CODE REVIEW

**Comment by skalish** [comment]
In this PR, there is a language mismatch between the `from_resource_id` and `by_name` functions.  I actually think "by" makes more sense when getting a resource from Tamr, whereas "from" fits de-serialization. So I might advocate for `by_name`, `by_resource_id`, and even `_by_url`, leaving only `_from_json` (and of course functions like `unified.from_project`).  This language is already being used in the docstrings for these functions: "Get dataset by URL" vs "Make dataset from JSON data (deserialize)".

**Comment by ianbakst** [comment]
I agree with this name change. This preposition makes more sense to me.

**Comment by skalish** [comment]
I've included the `from_resource_id` -> `by_resource_id` and `_from_url` -> `_by_url` renaming at the end of this PR, but I'm happy to roll back one or both if we decided to keep the old naming.

File: tests/tamr_client/fake_json/dataset/test_dataset/test_by_name.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+[
+    {
+        "request": {
+            "method": "GET",
+            "path": "datasets?filter=name==dataset 1 name"
```
Comment (pcattori) [inline tests/tamr_client/fake_json/dataset/test_dataset/test_by_name.json:5]: nit: Ideally, tests should be as realistic as possible, so `dataset 1 name` should be replaced with something like `Customers_2019.csv` or something. We are inconsistent about this and I think we can wait to fix this until when we switch over to faking a test server.

**Pull request merged**